### ISSUE TITLE
JsonReaderException when parsing other JSON within scanValuesFromStream

### ISSUE THREAD

**Issue by ski309** [plokhotnyuk/jsoniter-scala#923]
Hello, I just discovered that if I call `scanJsonValuesFromStream` to parse a file of JSON objects, then try to parse separate JSON from within the lambda, a JsonReaderException is thrown with the message, `expected '{', offset: 0x00000000, buf: ...`. The buffer data shown is in a location within the InputStream.

example:
```
val aData: String = ???
val in: InputStream = ???
Using.resource(in) { in =>
  scanJsonValuesFromStream[A](in) { a =>
    readFromString[A](aData)
  }
}
```

**Comment by plokhotnyuk** [comment]
Hi, @ski309 !

All `readFrom...` and `writeTo...` methods aren't re-entrant due to using the same `JsonReader` or `JsonWriter` instances shared through the thread local cache except `readFromStringReentrant` and `writeToStringReentrant`.

Please try re-entrant ones for your case.

**Comment by ski309** [comment]
I see... I hadn't noticed those methods. Are there any plans to introduce reentrant methods for byte arrays or input streams? Also, are there any plans to allow access to create our own `JsonReader` or `JsonWriter` instances?

**Comment by ski309** [comment]
In any case, thank you for your help!

**Comment by plokhotnyuk** [comment]
@ski309 A good option for the `scanValuesFromStream` callback would be passing a parsed message to different thread for handling of parsed message. As example through some blocking queue with limited size for back pressure. What do you think about that?

**Comment by ski309** [comment]
@plokhotnyuk I think having to pass messages to a different thread would reduce performance in a hot path in my code. For now, I've switched the JSON parsing within the callback back to Jackson, as it was beforehand.

In the future, I might remove the call to `scanValuesFromStream` entirely, and manually read the stream, and parse objects from the stream via jsoniter an object at a time.

One other suggestion would be to mention in the `JsonReaderException` that the error might have been caused by non-reentrant code?

Thanks for your help!
